### PR TITLE
Add smoke test, in order to check bundle integration

### PR DESCRIPTION
**Pull request by TBoileau**
* Delete some dead code
* Rename `Mock` folder to `Dummy`
* Rename `Functionnal` to `Functional`
* Move tests in `Functional` to `Unit` folder
* Install `symfony/phpunit-bridge`
* Implement a smoke test `AirtableClientTest`, in order to check the good integration in Symfony
* Implement `AirtableClientInterface` to respect **Dependency Inversion Principle**
* Add some `declare(strict_types=1);`
* Make services `public` in order to test its in smoke test
* Refactor `AirtableClient` by adding return type to all methods present in `AirtableClientInterface`
* Update `README.md`
* Add more assertions in `AirtableClientTest`
* Add test if the method `findTheLatest` return null
* Code coverage at 100%
* Add test in continuous integration by adding a github action

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -29,9 +29,12 @@
     }
   },
   "require-dev": {
+    "ext-json": "*",
     "phpunit/phpunit": "^9.5",
```
Comment (Spomky) [inline composer.json]: Now that `symfony/phpunit-bridge` is used and `./vendor/bin/simple-phpunit` is available, is it relevant to keep phpunit as a dev dep?

Comment (TBoileau) [reply composer.json]: We can actually delete phpunit/phpunit dependency ;) 
I'll take care of it

File: tests/Unit/AirtableClientTest.php [hunk 2]
```diff
@@ -213,8 +213,11 @@ public function findOneById_will_return_object_if_data_class_is_set()
         static::assertInstanceOf(Customer::class, $results->getFields());
     }
 
-    private function createHttpClientMock(string $expectedCallUrl, array $expectedJsonData = [], string $expectedMethod = 'GET')
-    {
+    private function createHttpClientMock(
+        string $expectedCallUrl,
+        array $expectedJsonData = [],
+        string $expectedMethod = 'GET'
+    ): HttpClientInterface {
```
Comment (liorchamla) [inline tests/Unit/AirtableClientTest.php:312]: C'est marrant, dans AirtableClient tu as remis sur une seule ligne le construct, et ici pour cette méthode tu as mis sur 3 lignes :p

**Review by liorchamla — APPROVED** [review]
Everything is more beautiful when @TBoileau is coding :p

File: composer.json (hunk 1 above)
Comment (liorchamla) [reply composer.json]: Will we be able to keep using Pest with the simple-phpunit ? What about the watcher ?

File: tests/Unit/AirtableClientTest.php (hunk 2 above)
Comment (TBoileau) [reply tests/Unit/AirtableClientTest.php:312]: Ah non mais je me suis fait feinter par le rebase :D

Comment (yoanbernabeu) [reply tests/Unit/AirtableClientTest.php:312]: La CI veillera au grain ;-)

File: composer.json (hunk 1 above)
Comment (TBoileau) [reply composer.json]: Normally, You shouldn't install this kind of dependency (like the watcher or phpunit), because it's specific to each developper

File: src/Resources/config/services.php [hunk 3]
```diff
@@ -2,29 +2,23 @@
...
 
-    $container->set(AirtableClient::class)
+    $container->set(AirtableClient::class, AirtableClient::class)
         ->args([
             '%yoanbernabeu_airtable_client.airtable_client.key%',
             '%yoanbernabeu_airtable_client.airtable_client.id%',
         ])
     ;
+
+    $container->alias(AirtableClientInterface::class, AirtableClient::class);
```
Comment (ismail1432) [inline src/Resources/config/services.php]: :bulb: `$container` is fluent,
```php
$container->set()
->alias()
;
```

IMO having a "key" instead of FQCN as an alias is more easier to manipulate for the end user (definition, decoration...)
`->alias(AirtableClientInterface::class, 'airtable_client')` but don't forget to create `airtable_client`

File: tests/Functional/AirtableClientKernel.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+namespace Yoanbernabeu\AirtableClientBundle\Tests\Functional;
+
+use Symfony\Bundle\FrameworkBundle\FrameworkBundle;
+use Symfony\Bundle\FrameworkBundle\Kernel\MicroKernelTrait;
+use Symfony\Component\Config\Loader\LoaderInterface;
+use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
+use Symfony\Component\HttpKernel\Kernel;
+use Yoanbernabeu\AirtableClientBundle\AirtableClientBundle;
+
+class AirtableClientKernel extends Kernel
```
Comment (ismail1432) [inline tests/Functional/AirtableClientKernel.php:14]: :+1:

File: .github/workflows/testing.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# .github/workflows/testing.yml
```
Comment (ismail1432) [inline .github/workflows/testing.yml:1]: :ok_hand:

File: src/Resources/config/services.php (hunk 3 above)
Comment (TBoileau) [reply src/Resources/config/services.php]:
```php
$container->set('airtable_client', AirtableClient::class)
        ->args([
            '%yoanbernabeu_airtable_client.airtable_client.key%',
            '%yoanbernabeu_airtable_client.airtable_client.id%',
            service(HttpClientInterface::class),
            service(ObjectNormalizer::class),
        ])
    ;

    $container->alias(AirtableClientInterface::class, AirtableClient::class);
```

Is not the same that : 
```php
    $container->set('airtable_client', AirtableClient::class)
        ->args([
            '%yoanbernabeu_airtable_client.airtable_client.key%',
            '%yoanbernabeu_airtable_client.airtable_client.id%',
            service(HttpClientInterface::class),
            service(ObjectNormalizer::class),
        ])
        ->alias(AirtableClientInterface::class, AirtableClient::class)
    ;
```

The object return by the method `args` is not the `$container`

Comment (liorchamla) [reply src/Resources/config/services.php]: Btw u don't "have to" add services references to HttpClientInterface or ObjectNormalizer since it is autowirid (i think ?)

File: .github/workflows/testing.yml (hunk 5 above)
Comment (TBoileau) [reply .github/workflows/testing.yml:1]: I don't konw why it's not working. Not recognized by Github...

File: tests/Unit/AirtableClientTest.php (hunk 2 above)
Comment (TBoileau) [reply tests/Unit/AirtableClientTest.php:312]: @yoanbernabeu j'ai ajouté une Github action pour mettre en place les tests dans la CI

File: src/Resources/config/services.php (hunk 3 above)
Comment (TBoileau) [reply src/Resources/config/services.php]: I let the code of rebase #5, I don't even seen this 2 lines of code, [I delete them and we'll see](https://github.com/yoanbernabeu/Airtable-Client-Bundle/pull/6/commits/98ea02423449265420b78e1be1433e615d4d3951)

Comment (Spomky) [reply src/Resources/config/services.php]: What is the result? I had in mind that either we don't define the arguments (and then autowiring is used) or we have to define everything.
I may be wrong, but because of the presence of the 2 params, other arguments shall be set.

Comment (ismail1432) [reply src/Resources/config/services.php]: `args` comes from `ArgumentTrait` which is used by `ServiceConfigurator` so it should works too but it's also OK like you did :sweat_smile:

Comment (TBoileau) [reply src/Resources/config/services.php]: It works with and without the references ;)

Comment (ismail1432) [reply src/Resources/config/services.php]: >I may be wrong, but because of the presence of the 2 params, other arguments shall be set.

Not sure about this, I think missing params will be autowired

Comment (TBoileau) [reply src/Resources/config/services.php]: > `args` comes from `ArgumentTrait` which is used by `ServiceConfigurator` so it should works too but it's also OK like you did 😅

Yes it's 2 differents instance of `ServiceConfigurator`, but that make any difference, you right, it works too, so I change it for a fluent one

File: composer.json (hunk 1 above)
Comment (TBoileau) [reply composer.json]: @yoanbernabeu I am confused about the use of pestphp and phpunit-watcher, I don't think they are essential to test you bundle. WHat do you think ?

Comment (yoanbernabeu) [reply composer.json]: > @yoanbernabeu I am confused about the use of pestphp and phpunit-watcher, I don't think they are essential to test you bundle. WHat do you think ?

Yes, let's keep it simple?

Comment (TBoileau) [reply composer.json]: So I can delete them from `composer.json` ?

Comment (yoanbernabeu) [reply composer.json]: OK for me

Comment (TBoileau) [reply composer.json]: Comitted, you can resolve this conversation if you want ;)

File: src/AirtableRecord.php [hunk 6]
```diff
@@ -93,21 +95,4 @@ public function getFields()
...
     }
-
-    public function setFields($fields): self
-    {
-        $this->fields = $fields;
-
-        return $this;
-    }
-
-    public function getId(): string
```
Comment (liorchamla) [inline src/AirtableRecord.php:99]: Why did we remove those getters ? The whole point of having an AirtableRecord is to have autocompletion and improve DX to retrieve those values imo :x

File: tests/Unit/Dummy/Customer.php [hunk 7]
```diff
@@ -9,4 +9,5 @@ class Customer
     public int $id;
     public string $firstName;
     public string $lastName;
+    public string $birthDay;
```
Comment (liorchamla) [inline tests/Unit/Dummy/Customer.php:12]: We do not need this property here, I just wanted to test that we would not have an exception if the result of Airtable contains fields that does not exist in the DTO class :x

**Review by liorchamla — CHANGES_REQUESTED** [review]

File: tests/Unit/Dummy/Customer.php (hunk 7 above)
Comment (TBoileau) [reply tests/Unit/Dummy/Customer.php:12]: Yes but this field is used in the request sent to Airtable : 
````php
        // Setup the dummy HttpClient
        $httpClient = $this->createHttpClientMock(
            "https://api.airtable.com/v0/MOCK_ID/MOCK_TABLE",
            [
                "records" => [
                    [
                        "id" => "MOCK_ID",
                        "fields" => [
                            "firstName" => "MOCK_FIRST_NAME",
                            "lastName" => "MOCK_LAST_NAME",
                            "id" => 1,
                            "birthDay" => "1986-10-30"
                        ],
                        "createdTime" => '2021-05-20T20:05:01.000Z'
                    ]
                ]
            ]
        );
```

File: src/AirtableRecord.php (hunk 6 above)
Comment (TBoileau) [reply src/AirtableRecord.php:99]: It's dead code, it's used anywhere, I athink like a PC, `it's not used, so it's unnecessary`. 

About DX of AirtableRecord, it's "poor" (we need to specify in phpdoc the return type of `$record->getFields()`, like [here](https://github.com/TBoileau/Airtable-Client-Bundle/blob/test-sf-integration/tests/Unit/AirtableClientTest.php#L65)), Don't you want to set up some Data Mapping ? :p

Comment (TBoileau) [reply src/AirtableRecord.php:99]: @liorchamla I put the deleted methods back into AirtableRecord ;) and add some tests

File: src/Resources/config/services.php (hunk 3 above)
Comment (TBoileau) [reply src/Resources/config/services.php]: Can we resolve this conversation ? @ismail1432

File: tests/Unit/Dummy/Customer.php (hunk 7 above)
Comment (liorchamla) [reply tests/Unit/Dummy/Customer.php:12]: It was a test to see if the ObjectNormalizer would throw when it finds a birthDay field that does not exist in the DTO class :) You could delete the birthDay field in the test right here more than add a birthDay property inside Customer::class ^^

Or just leave everything as it is :D

**Review by liorchamla — APPROVED** [review]

File: tests/Unit/Dummy/Customer.php (hunk 7 above)
Comment (TBoileau) [reply tests/Unit/Dummy/Customer.php:12]: We can let  `brithday`, just a another data like `firstName`.

Comment (liorchamla) [reply tests/Unit/Dummy/Customer.php:12]: Really not that important anyway lets resolve this and go on :)

File: src/AirtableRecord.php (hunk 6 above)
Comment (TBoileau) [reply src/AirtableRecord.php:99]: @liorchamla the presence of `setFields` method implicitly means that an AirtableRecord is not immutable. In my opinion, the object must be immutable, so I removed the "setFields" method. What do you think?

Comment (liorchamla) [reply src/AirtableRecord.php:99]: 100% ok :) @ismail1432 already pointed this out :)

**Review by liorchamla — APPROVED** [review]

**Comment by yoanbernabeu** [comment]
This PR is awesome my friends.
Thank you for your many contributions!

**Review by yoanbernabeu — APPROVED** [review]

**Review by liorchamla — APPROVED** [review]

**Comment by yoanbernabeu** [comment]
Merge after the live Twitch of @TBoileau !
Thank you all !

**Comment by TBoileau** [comment]
> Merge after the live Twitch of @TBoileau !
> Thank you all !

Perfect

**Review by liorchamla — APPROVED** [review]

**Comment by liorchamla** [comment]
Bravo à tous c'était du beau boulot !
